Views management refactoring

ixrock opened the pull request:
Back to managing views with `webview` (iframe) in favour of smoothest clusters switching

nevalla committed:
Fix CRD api parsing (#622)

Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

Nokel81 committed:
Fix Resource Quota Rendering (#624)

* add parsing of quota values which are non-numeric in the general case

* add unit tests

Signed-off-by: Sebastian Malton <smalton@mirantis.com>

Co-authored-by: Sebastian Malton <smalton@mirantis.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
part 1

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
part 2

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
cluster-view battling -- part 1 (iframe failing support nodeintegration)

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
ipc-refactoring, fixes

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
cluster-view battling -- part 2

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
cluster-view battling -- part 3

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
fixes

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
fix menu navigation

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

aleksfront committed:
Fixing Cluster Settings layout (#651)

* A bit of cleaning in Add Cluster page

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Adding head-col to WizardLayout

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Fixing Cluster Settings general layout bugs

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Cluster Status view refactoring

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Install Metrics component refactoring

Using notifications for error, removed picking
button icon method, simplified button generation.

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Remove icons / checks from RemoveClusterButton

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Fixing colorError in Input styles

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Preventing Input's spellchecking

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* ClusterNameSettings refactoring

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* ClusterWorkspaceSettings refactoring/fixing

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* ClusterProxySetting refactoring

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* ClusterPrometheusSetting refactoring

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Clean up Removal section

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Glued InstallMetrics & InstallUserMode into 1 component

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Removing unused styles in Cluster Settings

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Cluster Settings styling

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Adding close button to settings header

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* ClusterHomeDirSetting refactoring

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* FilePicker restyling

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Fixing Prometheus selector

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Fixing Hashicon

Passing cluster name instead of cluster id to prevent
icon changing while typing new
cluster name

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Minor ClusterSettings fixes

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Increasing opacity for non-interactive icons

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Keep feature install loading state

Waiting for props to change before
disabling loading state (gray button
width spinner)

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Remove arrays in disposeOnUnmount()

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Fix Cluster select behavior

Now clicking cluster icon in sidebar
always leads to / dashboard. And
'Settings' submenu switches active
cluster at first and only the showing
Cluster Settings

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Using structuralComparator in feature installer

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Saving input fields on blur

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Setting Select color same as Input color

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

ixrock committed:
cluster-view bugfixing, attempt 100500..

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
base-migration-branch merge-fix

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

jim-docker committed:
adding port-forward for containers in pods (#528)

* adding port-forward for containers in pods

address review comments

use more idiomatic approach for async code

move some files in advance of merge conflict with Lens restructure work

* Separate the port forward links in the UI (so they don't all spin when one link is clicked)

* minor fixes

* addressed review comments (replaced <p> with <div>, moved key attribute to proper element)

* fix lint issue

* removed extraneous <div> from pod container port details

Signed-off-by: Jim Ehrismann <jehrismann@mirantis.com>

Nokel81 committed:
use the Kubernetes regex for matching system names (#659)

* use the Kubernetes regex for matching system names

Signed-off-by: Sebastian Malton <smalton@mirantis.com>

Co-authored-by: Sebastian Malton <smalton@mirantis.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
cluster-view reworks

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
todo/fixme: hide active view on disconnect

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

aleksfront committed:
Removing legacy browserCheck() utility

Signed-off-by: alexfront <alex.andreev.email@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

aleksfront committed:
Revert select box-shadow (outline) color

Signed-off-by: alexfront <alex.andreev.email@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

aleksfront committed:
Allowing to save prometheus service address

Signed-off-by: alexfront <alex.andreev.email@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
moved some runtime deps from dev-deps-list in package.json

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

aleksfront committed:
Improving code readabilty

Signed-off-by: alexfront <alex.andreev.email@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
cluster-view fixes

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

aleksfront committed:
Setting overflow:hidden on lens-view elem (#664)

This prevents scrollbar flickering caused by Animation component
with 'slide-left' appearing.

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

ixrock committed:
cluster-view more fixes & UX optimizations

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

aleksfront committed:
Fixing broken/missing icons (#665)

* Normalizing titles size and weight

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Replacing deployment scale icon

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Replacing download logs icon

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Making <h4> tag bigger than <h5>

Signed-off-by: alexfront <alex.andreev.email@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
fix: cluster-manager initial loading, redirect to startPage from any "404"-route

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
fix: handle navigation from global menu

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
more fixes

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
fix menu/navigation

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
moar random fixes

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

jakolehm commented on src/main/index.ts:
```diff
@@ -41,8 +40,7 @@ async function main() {
   const updater = new AppUpdater()
   updater.start();
 
-  registerFileProtocol(appProto, app.getPath("userData"));
```
Why `appProto` has been removed?

ixrock replied:
Because it's never used

jakolehm replied:
Might be a dumb question but how do we display uploaded cluster icons then?

Nokel81 replied:
I switched to saving the base64 encoding of them in the user store

Nokel81 replied:
It removes the filename issue that we were seeing with some unicode characters, and it removes the security problems when running on dev mode.

jakolehm replied:
> I switched to saving the base64 encoding of them in the user store

Ok, do we have a migration for that?

Nokel81 replied:
I did not code one up, I should do that then

jakolehm replied:
> I did not code one up, I should do that then

Yes we should have a migration. Otherwise Lens will drop all the existing cluster icons on upgrade.

Nokel81 commented on src/common/cluster-store.ts:
```diff
@@ -64,11 +62,13 @@ export class ClusterStore extends BaseStore<ClusterStoreModel> {
+          const cluster = this.getById(state.id);
+          if (cluster) {
+            cluster.updateModel(state)
+          }
```
```suggestion
          this.getById(state.id)?.updatedModel(state);
```

Nokel81 commented on src/common/ipc.ts:
```diff
@@ -2,94 +2,133 @@
+  error?: E;
+}
+
+export interface IpcChannelInit {
```
This doesn't seem to be about the initial channel, wouldn't something like `IpcChannelOpts` be more fitting?

ixrock replied:
yep, sounds good

Nokel81 commented on src/common/ipc.ts:
```diff
@@ -2,94 +2,133 @@
+}
+
+export function createIpcChannel({ autoBind = true, mode = IpcMode.ASYNC, timeout = 0, handle, channel }: IpcChannelInit) {
+  channel = `${mode}:${channel}`
```
Shouldn't it be an error to given `channel === ""`?

ixrock replied:
not critical imo, why not an empty string as a name?

Nokel81 replied:
Just seems like a weird that a channel called `"sync:"` or `"async:"` could be made. Another possible solution could be 

```javascript
channel = `${mode}:${channel || "default"}`
```

Nokel81 commented on src/common/ipc.ts:
```diff
@@ -2,94 +2,133 @@
+          }
+          if (mode === IpcMode.SYNC) {
+            event.returnValue = res;
+          }
```
shouldn't this either be a `switch` or `if ... else if`?

ixrock replied:
doesn't really matter, only if it's a question of code-readability? :)

nevalla replied:
IMHO `if ... else if` would be better here.

Nokel81 commented on src/common/ipc.ts:
```diff
@@ -2,94 +2,133 @@
+      if (mode === IpcMode.SYNC) {
+        ipcRenderer.sendSync(channel, req)
+      }
+      return new Promise(async (resolve, reject) => {
```
why are we returning a promise if the mode is `sync`?

ixrock replied:
maybe we should, haven't even tested `sync` mode and it's better to use this "as last resort" as defined in docs:
https://www.electronjs.org/docs/api/ipc-renderer#ipcrenderersendsyncchannel-args

nevalla replied:
Just thinking that do we need `sync` mode here if we are only using async messaging?

ixrock replied:
Well, this `IPC` helper is intended to be multi-functional, so I would keep it, but have to test if it's really works as expected.

Nokel81 commented on src/common/user-store.ts:
```diff
@@ -71,11 +71,12 @@ export class UserStore extends BaseStore<UserStoreModel> {
-        .filter(ctx => ctx.cluster)
-        .filter(ctx => !this.seenContexts.has(ctx.name))
-        .forEach(ctx => this.newContexts.add(ctx.name));
+      localContexts.forEach(({ cluster, name }) => {
```
why this change, the older version is more clear

ixrock replied:
Not for me, sorry. Older version more like functional-style, but I would not say it's more readable (for humans) + there is no need to run multiple cycles on what's can be updated in one run.

jakolehm replied:
I think this is matter of taste. I would revert this change as it's not related to views management at all, right?

Nokel81 commented on src/main/cluster.ts:
```diff
@@ -80,20 +84,15 @@ export class Cluster implements ClusterModel {
 
   @action
   async init(port: number) {
-    if (this.initialized) {
```
Is it now save to init twice?

ixrock replied:
No, init used only in single place in cluster-manager with that check

Nokel81 commented on src/main/lens-proxy.ts:
```diff
@@ -117,26 +119,17 @@ export class LensProxy {
+        return proxy.web(req, res, proxyTarget);
+      }
     }
+    this.router.route(cluster, req, res);
```
why route on an undefined cluster? Wouldn't it be better to change line 123 to be `if (!cluster) throw...` or `if (!cluster) return ...`?

Nokel81 commented on src/main/routes/metrics-route.ts:
```diff
@@ -1,52 +1,34 @@
-      metricsUrl = `${kubeProxyUrl}/api/v1/namespaces/${prometheusPath}/proxy${cluster.getPrometheusApiPrefix()}/api/v1/query_range`
-      prometheusProvider = await contextHandler.getPrometheusProvider()
+      prometheusPath = await cluster.contextHandler.getPrometheusPath()
+      prometheusProvider = await cluster.contextHandler.getPrometheusProvider()
```
Shouldn't these be `await`-ed in together?

ixrock replied:
Not critical, but why not

Nokel81 commented on src/main/window-manager.ts:
```diff
@@ -1,68 +1,80 @@
   protected windowState: windowStateKeeper.State;
 
-  constructor(protected proxyPort: number, showSplash = true) {
+  @observable activeClusterId: ClusterId;
```
If adding an observable field, shouldn't the class be marked as `observer`?

ixrock replied:
No, `@observer` is used only with React-components. If component use in `render()` some mobx-observables only then it must have this decorator (to be able to re-render view).

Nokel81 replied:
Ah thanks for the information, I did not know that

Nokel81 commented on src/renderer/bootstrap.tsx:
```diff
@@ -0,0 +1,38 @@
+  ]);
+
+  // init app's dependencies if any
+  if (App.init) {
+    await App.init();
+  }
```
```suggestion
  await App.init?.call();
```

ixrock replied:
Don't thinks it's a good idea imo..
`App.init?.call()` creates more cognitive load (since .call() rarely used in general or this way)

Nokel81 replied:
Makes sense

Nokel81 commented on src/renderer/components/+apps-releases/release.store.ts:
```diff
@@ -58,7 +58,7 @@ export class ReleaseStore extends ItemStore<HelmRelease> {
     let items;
     try {
-      const { isAdmin, allowedNamespaces } = getHostedCluster();
+      const { isAdmin, allowedNamespaces } = getHostedCluster()
```
why all the removing of semicolons?

ixrock replied:
we don't have strict rules on that :) in general semicolon redundant and should be avoid imo..

Nokel81 commented on src/renderer/components/cluster-manager/cluster-manager.tsx:
```diff
@@ -1,52 +1,116 @@
+  return !!lensViews.get(clusterId)?.isLoaded;
+}
+
+// todo: figure out how to replace <webview>-tag to <iframe> with nodeIntegration=true
```
Maybe or https://github.com/react-native-community/react-native-webview

ixrock replied:
it was more question "why <iframe> previously get to work with vue-app" 
p.s your link follows to `react-native-community`

Nokel81 reviewed (commented):
Looks good from a top level

ixrock committed:
clean up

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

aleksfront committed:
Fix Cluster Settings style specificity

Signed-off-by: alexfront <alex.andreev.email@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
fix: navigate to cluster-settings page from clusters-menu

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

nevalla committed:
Revert system menu layout (#671)

Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

jakolehm commented on src/main/cluster-manager.ts:
```diff
@@ -50,23 +42,8 @@ export class ClusterManager {
-    return cluster;
+    logger.debug(`getClusterForRequest(): ${req.headers.host}${req.url}`)
+    const clusterId = req.headers.host.split(".")[0]
+    return this.getCluster(clusterId)
```
Why logic has been removed? Don't we use `127.0.0.1:<port>/uuid` anymore anywhere?

ixrock replied:
Seems so.. Looks like outdated code from very beginning of lens-app.

jakolehm replied:
I asked because I `port/uuid` addresses were used when doing requests outside of renderer (wildcard dns to localhost works only within browser scope). Anyway I don't like to see changes that are totally outside of PR scope, it makes review unnecessary complicated.

ixrock replied:
Ok, reverted back. Someone should check if this is actually works, cause top comment:
`// lens-server is connecting to 127.0.0.1:<port>/<uid>`
which unclear in terms of what you saying how this might be used outside.

ixrock committed:
fix: navigating to cluster-view from common-view

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
revering user-store refactoring

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
moving cluster-settings view lens-app views (common)

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
reverted cluster-manager.ts/getClusterForRequest()

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

aleksfront committed:
Adding unit tests to ClusterStore (#678)

Signed-off-by: alexfront <alex.andreev.email@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

aleksfront committed:
Navigate to landing page on cluster disconnect

Signed-off-by: alexfront <alex.andreev.email@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

aleksfront committed:
Add Cluster page visual fixes

Signed-off-by: alexfront <alex.andreev.email@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

aleksfront committed:
Improve isUrl validator

Allowing to check http(s) and format
without domain like http://localhost:8080

Signed-off-by: alexfront <alex.andreev.email@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

aleksfront committed:
Fixing Cluster Settings typo

Signed-off-by: alexfront <alex.andreev.email@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
fix: use `env NODE_ENV=production` for `yarn compile`

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
prod-build fix

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
extracting missing localization

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

aleksfront committed:
Reconnecting cluster after selection (#685)

* Reconnecting selected cluster

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Cleaning up ClusteStatus render() method

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

ixrock committed:
fix: skip redefining global `__static` in tests

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

aleksfront committed:
Preferences section restyling (#700)

* Adding 'centering' option to WizardLayout

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Using centered layout in Cluster Settings

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Restyling Preferences section

Signed-off-by: alexfront <alex.andreev.email@gmail.com>

* Simplifying Helm loading state

Signed-off-by: alexfront <alex.andreev.email@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

ixrock committed:
Replace webview to iframe (#701)

* replace webview-tag to iframe with nodeIntegrationInSubFrames=true

Signed-off-by: Roman <ixrock@gmail.com>

* clean up

Signed-off-by: Roman <ixrock@gmail.com>

* chore: fix type ipc#handle

Signed-off-by: Roman <ixrock@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

nevalla committed:
Fix integration tests (#698)

Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

aleksfront committed:
Minor WizardLayout fix

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>
Signed-off-by: Lauri Nevala <lauri.nevala@gmail.com>

nevalla committed:
Merge branch 'vue_react_migration' into views_management_refactoring
